mail_server_relay_disallowed/models/ir_mail_server.py

```diff
@@ -0,0 +1,132 @@
+# Copyright 2020 ForgeFlow, S.L.
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+import logging
+import smtplib
+import threading
+
+from odoo import api, models
+from odoo.tools import ustr
+from odoo.tools.translate import _
+
+from odoo.addons.base.models.ir_mail_server import (
+    MailDeliveryException,
+    extract_rfc2822_addresses,
+)
+
+_logger = logging.getLogger(__name__)
+_test_logger = logging.getLogger("odoo.tests")
+
+
+class IrMailServer(models.Model):
+    _inherit = "ir.mail_server"
+
+    NO_VALID_RECIPIENT = (
+        "At least one valid recipient address should be "
+        "specified for outgoing emails (To/Cc/Bcc)"
+    )
+
+    @api.model
+    def send_email(
+        self,
+        message,
+        mail_server_id=None,
+        smtp_server=None,
+        smtp_port=None,
+        smtp_user=None,
+        smtp_password=None,
+        smtp_encryption=None,
+        smtp_debug=False,
+        smtp_session=None,
+    ):
+        """Override the standard method to fix the issue of using a mail
+        client where relaying is disallowed."""
+        # Use the default bounce address **only if** no Return-Path was
+        # provided by caller.  Caller may be using Variable Envelope Return
+        # Path (VERP) to detect no-longer valid email addresses.
+        smtp_from = (
+            message["Return-Path"]
+            or self._get_default_bounce_address()
+            or message["From"]
+        )
+        assert (
+            smtp_from
+        ), "The Return-Path or From header is required for any outbound email"
```

Comment:
Thanks for your module.
It seems than assert expression shouldn't not be in production code like here
https://stackoverflow.com/a/5143044 because could be skipped according to the execution mode of your app instance. Then assert should be reserve for unit tests where there is no production mode in execution. 
But yes Odoo SA itself use this statement in its code not only for tests. As usual, OCA promotes the best practices before Odoo SA
https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement
There is a lot of documentation on this subject in python ecosystem
Not blocking for me, just for information

Reply:
yeah, but here this part of code is just copied from odoo (it's an override) and we want to change the minor possible the code, in order to be able to check easy what change has been done comparing to original code.